File: app/src/main/java/forpdateam/ru/forpda/views/messagepanel/advanced/adapters/PanelItemAdapter.java
package forpdateam.ru.forpda.views.messagepanel.advanced.adapters;

import android.graphics.Bitmap;
import android.graphics.Color;
import android.graphics.ColorFilter;
import android.graphics.PorterDuff;
import android.graphics.PorterDuffColorFilter;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.TextView;

import com.nostra13.universalimageloader.core.ImageLoader;
import com.nostra13.universalimageloader.core.listener.SimpleImageLoadingListener;

import java.util.Collections;
import java.util.List;

import forpdateam.ru.forpda.App;
import forpdateam.ru.forpda.R;
import forpdateam.ru.forpda.views.messagepanel.advanced.ButtonData;

/**
 * Created by radiationx on 08.01.17.
 */

public class PanelItemAdapter extends RecyclerView.Adapter<PanelItemAdapter.ViewHolder> implements ItemDragCallback.ItemTouchHelperAdapter {
    public final static int TYPE_ASSET = 0;
    public final static int TYPE_DRAWABLE = 1;
    private final ColorFilter colorFilter = new PorterDuffColorFilter(Color.BLACK, PorterDuff.Mode.SRC_ATOP);
    private List<ButtonData> items;
    private List<String> urlsToAssets;
    private int type = -1;

    public PanelItemAdapter(List<ButtonData> items, List<String> urlsToAssets, int type) {
        this.items = items;
        this.urlsToAssets = urlsToAssets;
        this.type = type;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.message_panel_advanced_item, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(final ViewHolder holder, final int position) {
        ButtonData item = items.get(position);
        if (type == TYPE_ASSET) {
            ImageLoader.getInstance().loadImage(urlsToAssets.get(position), new SimpleImageLoadingListener() {
                @Override
                public void onLoadingComplete(String imageUri, View view, Bitmap loadedImage) {
                    holder.button.setImageBitmap(loadedImage);
                }
            });
        } else if (type == TYPE_DRAWABLE) {
            holder.button.setImageDrawable(App.getAppDrawable(item.getIconRes()));
            holder.button.setColorFilter(colorFilter);
        }
        if (item.getTitle() == null) {
            holder.title.setVisibility(View.GONE);
        } else {
            holder.title.setText(item.getTitle());
            holder.title.setVisibility(View.VISIBLE);
        }
    }

    private OnItemClickListener itemClickListener;

    public interface OnItemClickListener {
        void onItemClick(ButtonData item);
    }

    public void setOnItemClickListener(final OnItemClickListener mItemClickListener) {
        this.itemClickListener = mItemClickListener;
    }

    @Override
    public int getItemCount() {
        return items.size();
    }

    public class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        public ImageButton button;
        public TextView title;

        public ViewHolder(View view) {
            super(view);
            view.setOnClickListener(this);
            button = (ImageButton) view.findViewById(R.id.item_icon);
            title = (TextView) view.findViewById(R.id.item_title);
        }

        @Override
        public void onClick(View v) {
            if (itemClickListener != null) {
                itemClickListener.onItemClick(items.get(getLayoutPosition()));
            }
        }
    }

    @Override
    public void onItemDismiss(int position) {
        items.remove(position);
        notifyItemRemoved(position);
    }

    @Override
    public void onItemMove(int fromPosition, int toPosition) {
        if (fromPosition < toPosition) {
            for (int i = fromPosition; i < toPosition; i++) {
                Collections.swap(items, i, i + 1);
            }
        } else {
            for (int i = fromPosition; i > toPosition; i--) {
                Collections.swap(items, i, i - 1);
            }
        }
        notifyItemMoved(fromPosition, toPosition);
        Log.e("SUKA", "onItemMove");
    }

}

File: app/src/main/java/forpdateam/ru/forpda/ui/views/messagepanel/advanced/adapters/PanelItemAdapter.kt
package forpdateam.ru.forpda.ui.views.messagepanel.advanced.adapters

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.view.isVisible
import androidx.recyclerview.widget.RecyclerView
import by.kirich1409.viewbindingdelegate.viewBinding
import com.nostra13.universalimageloader.core.ImageLoader
import forpdateam.ru.forpda.R
import forpdateam.ru.forpda.databinding.MessagePanelAdvancedItemBinding
import forpdateam.ru.forpda.extensions.setTintColor
import forpdateam.ru.forpda.ui.views.messagepanel.advanced.PanelListItem
import forpdateam.ru.forpda.ui.views.messagepanel.advanced.adapters.ItemDragCallback.ItemTouchHelperAdapter
import java.util.Collections

/**
 * Created by radiationx on 08.01.17.
 */
class PanelItemAdapter(
    private val items: MutableList<PanelListItem>,
    private val clickListener: (PanelListItem) -> Unit
) : RecyclerView.Adapter<RecyclerView.ViewHolder>(), ItemTouchHelperAdapter {

    override fun getItemViewType(position: Int): Int {
        return items[position]::class.hashCode()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val view = LayoutInflater.from(parent.context)
            .inflate(R.layout.message_panel_advanced_item, parent, false)
        return when (viewType) {
            PanelListItem.Smile::class.hashCode() -> SmileViewHolder(clickListener, view)
            PanelListItem.BBCode::class.hashCode() -> BBCodeViewHolder(clickListener, view)
            PanelListItem.Color::class.hashCode() -> ColorViewHolder(clickListener, view)
            else -> error("Unknown viewtype")
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        when (val item = items[position]) {
            is PanelListItem.Smile -> (holder as SmileViewHolder).bind(item)
            is PanelListItem.BBCode -> (holder as BBCodeViewHolder).bind(item)
            is PanelListItem.Color -> (holder as ColorViewHolder).bind(item)
        }
    }

    override fun getItemCount(): Int {
        return items.size
    }

    class SmileViewHolder(
        private val clickListener: (PanelListItem.Smile) -> Unit,
        view: View
    ) : RecyclerView.ViewHolder(view) {

        private val binding by viewBinding<MessagePanelAdvancedItemBinding>()

        fun bind(item: PanelListItem.Smile) {
            val assetUrl = "assets://smiles/${item.assetFileName}"
            ImageLoader.getInstance().cancelDisplayTask(binding.itemIcon)
            ImageLoader.getInstance().displayImage(assetUrl, binding.itemIcon)
            binding.itemTitle.visibility = View.GONE
            binding.root.contentDescription = item.text
            binding.root.setOnClickListener {
                clickListener.invoke(item)
            }
        }
    }

    class BBCodeViewHolder(
        private val clickListener: (PanelListItem.BBCode) -> Unit,
        view: View
    ) : RecyclerView.ViewHolder(view) {

        private val binding by viewBinding<MessagePanelAdvancedItemBinding>()

        fun bind(item: PanelListItem.BBCode) {
            binding.itemIcon.setImageResource(item.iconRes)
            val title = binding.itemTitle.context.getString(item.titleRes)
            binding.itemTitle.isVisible = true
            binding.itemTitle.text = title
            binding.root.contentDescription = title
            binding.root.setOnClickListener {
                clickListener.invoke(item)
            }
        }
    }

    class ColorViewHolder(
        private val clickListener: (PanelListItem.Color) -> Unit,
        view: View
    ) : RecyclerView.ViewHolder(view) {

        private val binding by viewBinding<MessagePanelAdvancedItemBinding>()

        fun bind(item: PanelListItem.Color) {
            binding.itemIcon.setImageResource(R.drawable.bg_circle_black)
            binding.itemIcon.setTintColor(item.color)
            binding.root.contentDescription = item.hexColor
            binding.itemTitle.isVisible = false
            binding.root.setOnClickListener {
                clickListener.invoke(item)
            }
        }
    }

    override fun onItemDismiss(position: Int) {
        items.removeAt(position)
        notifyItemRemoved(position)
    }

    override fun onItemMove(fromPosition: Int, toPosition: Int) {
        if (fromPosition < toPosition) {
            for (i in fromPosition until toPosition) {
                Collections.swap(items, i, i + 1)
            }
        } else {
            for (i in fromPosition downTo toPosition + 1) {
                Collections.swap(items, i, i - 1)
            }
        }
        notifyItemMoved(fromPosition, toPosition)
        //Log.d("FORPDA_LOG", "onItemMove");
    }
}
